standard/patterns.md

```diff
@@ -157,13 +169,233 @@ var_pattern
+
+### §positional-pattern-new-clause Positional pattern
+
+A *positional_pattern* checks that the input value is not `null`, invokes an appropriate `Deconstruct` method ([§12.7](expressions.md#127-deconstruction)), and performs further pattern matching on the resulting values.  It also supports a tuple-like pattern syntax (without the type being provided) when the type of the input value is the same as the type containing `Deconstruct`, or if the type of the input value is a tuple type, or if the type of the input value is `object` or `System.ITuple` and the runtime type of the expression implements `System.ITuple`.
+
+```ANTLR
+positional_pattern
+    : type? '(' subpatterns? ')' property_subpattern? simple_designation?
+    ;
+subpatterns
+    : subpattern (',' subpatterns)?
+    ;
+subpattern
+    : pattern
+    | identifier ':' pattern
+    ;
+```
+
+If *type* is omitted, the type is assumed to be the static type of the input value.
+Given a match of an input value to the pattern *type* `(` *subpatterns* `)`, a method is selected by searching in *type* for accessible declarations of `Deconstruct` and selecting one among them using the same rules as for the deconstruction declaration.
+It is an error if a *positional_pattern* omits the type, has a single *subpattern* without an *identifier*, has no *property_subpattern* and has no *simple_designation*. This disambiguates between a *constant_pattern* that is parenthesized and a *positional_pattern*.
+In order to extract the values to match against the patterns in the list,
+
+- If *type* is omitted and the input value's type is a tuple type, then the number of subpatterns shall to be the same as the cardinality of the tuple. Each tuple element is matched against the corresponding *subpattern*, and the match succeeds if all of these succeed. If any *subpattern* has an *identifier*, then that shall name a tuple element at the corresponding position in the tuple type.
+- Otherwise, if a suitable `Deconstruct` exists as a member of *type*, it is a compile-time error if the type of the input value is not pattern-compatible with *type*. At runtime the input value is tested against *type*. If this fails, then the positional pattern match fails. If it succeeds, the input value is converted to this type and `Deconstruct` is invoked with fresh compiler-generated variables to receive the output parameters. Each value that was received is matched against the corresponding *subpattern*, and the match succeeds if all of these succeed. If any *subpattern* has an *identifier*, then that shall name a parameter at the corresponding position of `Deconstruct`.
+- Otherwise, if *type* is omitted, and the input value is of type `object`, `System.ITuple`, or some type that can be converted to `System.ITuple` by an implicit reference conversion, and no *identifier* appears among the subpatterns, then the match uses `System.ITuple`.
+- Otherwise, the pattern is a compile-time error.
```

Comment:
Patterns (including property patterns and positional patterns) implicitly unwrap `System.Nullable<T>`, as part of their "checks that the input value is not `null`". Do we need to say something to this effect here?

Example:

```cs
class C
{
    void M(Point? p)
    {
        if (p is (var x, var y)) // True
        {
        }
    }
}

struct Point
{
    public int X, Y;
    public void Deconstruct(out int x, out int y) { x = X; y = Y; }
}
```

Reply:
Nothing needs to be said, that's how checking whether a `Nullable<T>` is `null` works, which is defined elsewhere in the Standard.